azext_iot/tests/dps/conftest.py

```diff
@@ -0,0 +1,264 @@
+def provisioned_iot_dps_no_hub_module(request) -> dict:
+    result = _iot_dps_provisioner(request)
+    yield result
+    if result:
+        _iot_dps_removal(result)
+
+
+def _iot_dps_provisioner(request, iot_hubs: Optional[List] = None) -> dict:
+    """Create a device provisioning service for testing purposes."""
+    dps_name = settings.env.azext_iot_testdps or generate_dps_id()
+    dps_list = cli.invoke(
+        'iot dps list -g "{}"'.format(ENTITY_RG)
+    ).as_json()
+
+    # Check if the generated name is already used
+    target_dps = None
+    for dps in dps_list:
+        if dps["name"] == dps_name:
+            target_dps = dps
+            logger.info(f"DPS {dps['name']} found.")
+            break
+
+    # Create the min version dps and assign the correct roles
+    if not target_dps:
+        if settings.env.azext_iot_testdps:
+            logger.error(f"DPS {dps_name} specified in pytest settings not found. DPS will be created")
+
+        base_command = f"iot dps create --name {dps_name} --resource-group {ENTITY_RG}"
+        if iot_hubs:
+            base_command += f" --tags hubname={iot_hubs[0]['name']}"
+        target_dps = cli.invoke(base_command).as_json()
+
+    assign_iot_dps_dataplane_rbac_role(target_dps)
+    sleep(10)
+
+    # Add link between dps and first iot hub
+    hub_host_name = None
+    if iot_hubs:
+        target_hub = iot_hubs[0]
+        linked_hubs = cli.invoke(
+            "iot dps linked-hub list --dps-name {} -g {}".format(dps_name, ENTITY_RG)
+        ).as_json()
+        hub_host_name = "{}.azure-devices.net".format(target_hub["name"])
+
+        if hub_host_name not in [hub["name"] for hub in linked_hubs]:
+            cli.invoke(
+                f"iot dps linked-hub create --dps-name {dps_name} -g {ENTITY_RG} "
+                f"--connection-string {target_hub['connectionString']}"
+            )
+
+    return {
+        "name": dps_name,
+        "resourceGroup": ENTITY_RG,
+        "dps": target_dps,
+        "connectionString": get_dps_cstring(dps_name, ENTITY_RG),
+        "hubHostName": hub_host_name,
+        "hubConnectionString": iot_hubs[0]["connectionString"] if iot_hubs else None,
+        "certificates": []
+    }
+
+
+def get_dps_cstring(dps_name: str, dps_rg: str, policy: str = "provisioningserviceowner") -> str:
+    return cli.invoke(
+        "iot dps connection-string show -n {} -g {} --policy-name {}".format(
+            dps_name, dps_rg, policy
+        )
+    ).as_json()["connectionString"]
+
+
+def _iot_dps_removal(dps):
+    for cert in dps["certificates"]:
+        if os.path.exists(cert):
+            try:
+                os.remove(cert)
+            except OSError as e:
+                logger.error(f"Failed to remove {cert}. {e}")
+    if not settings.env.azext_iot_testdps:
+        cli.invoke(
+            "iot dps delete --name {} --resource-group {}".format(
+                dps["name"], dps["resourceGroup"]
+            )
+        )
+
+
+# IoT Hub fixtures
```

Comment:
Q: should I:
- keep this generic so it can be merged with the other IoT Hub fixtures easily
- make it specific for DPS since only one hub needs to be created and add support for specifying a specific hub you want to use from pytest?

Reply:
Seems appropriate to me to keep the fixture definition (where scope is defined) specific for DPS, where-as the components used in implementation to carry out actions are generically defined and used. So a test can define the scope for which test infra will be evaluated however it doesn't need to re-invent the wheel for how an iot hub should be provisioned and configured.